File: project_test.m
%prueba proyecto
close all
clear ll
n = 100;                            %cantidad de datos
fvar = 0.1;                         %efecto del ruido
m = 8;                              %2,4,8,16,...
r = 5;                              %repeticiones (para comparar entre diferentes señales)

datos = round((m-1)*rand(r,n));     %datos
a1bar = zeros(1,r);
a1var = zeros(1,r);
for i=1:r
    angles = datos(i,:)*2*pi/m;              %fase correspondiente
    envolvente = (fvar*randn(1,n)+1).*exp(1i*(angles+fvar*randn(1,n)));  %envolvente, se agrega efectos aleatorios
    a = real(envolvente);           %parte real
    b = imag(envolvente);           %parte imaginaria

    %constelaciones
    lim = 1.5;
    figure(i)
    scatter(a,b)
    ylim([-lim,lim])
    xlim([-lim,lim])
    grid

    cond1 = (a<1.5 & a>0.5) & abs(b)<0.5;   %encontrar valores alrededor de 1 + j0

    %datos seleccionados
    a1 = a(cond1);
    b1 = b(cond1);

    figure(i+r)
    scatter(a1,b1)
    ylim([-lim,lim])
    xlim([-lim,lim])
    grid
    a1bar(i) = sum(a1)/size(a1,2);
    a1var(i) = sum((a1 - a1bar(i)).^2)/(size(a1,2)-1);
end

find(a1var==min(a1var))   %señal mas adecuada

File: project_test.py
import numpy as np
import math
import matplotlib.pyplot as plt

n=150
fvar=0.1
m=4            #numero de niveles de señal
r=1            #numero de señales para comparar entre
pi = math.pi
e=math.e
lim=1.5

datos = np.round((m-1)*np.random.rand(r,n))           #datos con m niveles
a1bar = np.zeros(r)
a1var = np.zeros(r)
b1bar = np.zeros(r)
b1var = np.zeros(r)
datos_desv_estandar = np.zeros(r)
for i in range(0,r):
    angles = datos[i,:]*2*pi/m
    mag = fvar*np.random.normal(0,1,n) + 1                  #magnitud, se agrega 
    a = mag*np.cos(angles+fvar*np.random.normal(0,1,n))   #efecto de ruido
    b = mag*np.sin(angles+fvar*np.random.normal(0,1,n))   #parte real e imaginaria

    plt.plot(a,b,'ro')
    plt.axis([-lim,lim,-lim,lim])
    plt.show()

    a1 = a[(0.5 < a) & (a < 1.5) & (b<0.5) & (b>-0.5)]    #sección 
    b1 = b[(0.5 < a) & (a < 1.5) & (b<0.5) & (b>-0.5)]

    plt.plot(a1,b1,'ro')
    plt.axis([-lim,lim,-lim,lim])
    plt.show()

    a1bar[i] = np.sum(a1)/a1.shape[0]                        #media y varianza de
    a1var[i] = np.sum((a1 - a1bar[i])**2)/(a1.shape[0]-1);      #los reales (a)
    b1bar[i] = np.sum(a1)/a1.shape[0]                        
    b1var[i] = np.sum((a1 - a1bar[i])**2)/(a1.shape[0]-1);
    datos_desv_estandar[i] = np.sqrt(a1var[i] + b1var[i])
    print(b1var)
    print(a1var)
    print(datos_desv_estandar)


resultado = np.where(datos_desv_estandar == np.amin(datos_desv_estandar))
 

print('Mejor señal para muestra es la señal: ', resultado[0])
print('Datos de la señal: ', datos[resultado,:])


















